test/unit/org/apache/cassandra/io/filesystem/ListenableFileSystem.java

```diff
@@ -0,0 +1,850 @@
+                                                                       onPrePosition, onPostPosition);
+    private final ListenableFileSystemProvider provider;
+
+    public ListenableFileSystem(FileSystem delegate)
+    {
+        super(delegate);
+        this.provider = new ListenableFileSystemProvider(super.provider());
+    }
+
+    public Unsubscribable listen(Listener listener)
+    {
+        boolean match = false;
+        if (listener instanceof OnPreOpen)
+        {
+            onPreOpen.add((OnPreOpen) listener);
+            match = true;
+        }
+        if (listener instanceof OnPostOpen)
+        {
+            onPostOpen.add((OnPostOpen) listener);
+            match = true;
+        }
+        if (listener instanceof OnPreRead)
+        {
+            onPreRead.add((OnPreRead) listener);
+            match = true;
+        }
+        if (listener instanceof OnPostRead)
+        {
+            onPostRead.add((OnPostRead) listener);
+            match = true;
+        }
+        if (listener instanceof OnPreTransferTo)
+        {
+            onPreTransferTo.add((OnPreTransferTo) listener);
+            match = true;
+        }
+        if (listener instanceof OnPostTransferTo)
+        {
+            onPostTransferTo.add((OnPostTransferTo) listener);
+            match = true;
+        }
+        if (listener instanceof OnPreWrite)
+        {
+            onPreWrite.add((OnPreWrite) listener);
+            match = true;
+        }
+        if (listener instanceof OnPostWrite)
+        {
+            onPostWrite.add((OnPostWrite) listener);
+            match = true;
+        }
+        if (listener instanceof OnPreTransferFrom)
+        {
+            onPreTransferFrom.add((OnPreTransferFrom) listener);
+            match = true;
+        }
+        if (listener instanceof OnPostTransferFrom)
+        {
+            onPostTransferFrom.add((OnPostTransferFrom) listener);
+            match = true;
+        }
+        if (listener instanceof OnPreForce)
+        {
+            onPreForce.add((OnPreForce) listener);
+            match = true;
+        }
+        if (listener instanceof OnPostForce)
+        {
+            onPostForce.add((OnPostForce) listener);
+            match = true;
+        }
+        if (listener instanceof OnPreTruncate)
+        {
+            onPreTruncate.add((OnPreTruncate) listener);
+            match = true;
+        }
+        if (listener instanceof OnPostTruncate)
+        {
+            onPostTruncate.add((OnPostTruncate) listener);
+            match = true;
+        }
+        if (listener instanceof OnPrePosition)
+        {
+            onPrePosition.add((OnPrePosition) listener);
+            match = true;
+        }
+        if (listener instanceof OnPostPosition)
+        {
+            onPostPosition.add((OnPostPosition) listener);
+            match = true;
+        }
+        if (!match)
```

Comment:
nit: What if, instead of tracking a boolean here, we just track which list (ex. `List<? extends Listener> match`) we add to, and then have the closure remove specifically from that list, instead of trying to remove it from everything? (i.e. `match` is null when there's no match above, an we throw and do nothing)